py/DREAM/Settings/Equations/IonSpecies.py

```diff
@@ -44,18 +44,23 @@
 IONIZATION_MODE_KINETIC = 2
 IONIZATION_MODE_KINETIC_APPROX_JAC=3
 
+ION_OPACITY_MODE_TRANSPARENT = 1
+ION_OPACITY_MODE_GROUND_STATE_OPAQUE = 2
+
 class IonSpecies:
     
-    def __init__(self, settings, name, Z, ttype=0, Z0=None, T=None, n=None, r=None, t=None, interpr=None, interpt=None, tritium=False):
+    def __init__(self, settings, name, Z, ttype=0, Z0=None, isotope=0, SPIMolarFraction=-1.0, opacity_mode = ION_OPACITY_MODE_TRANSPARENT, T=None, n=None, r=None, t=None, interpr=None, interpt=None, tritium=False):
         """
         Constructor.
 
         :param DREAMSettings settings: Parent DREAMSettings object.
         :param str name:               Name by which the ion species will be referred to.
         :param int Z:                  Ion charge number.
+        :param int isotope:            Ion mass number.
         :param int ttype:              Method to use for evolving ions in time.
         :param int Z0:                 Charge state to populate with given density.
         :param float n:                Ion density (can be either a scalar, 1D array or 2D array, depending on the other input parameters)
+        :param float SPIMolarFraction: Molar fraction of the SPI injection (if any). A negative value means that this species is not part of the SPI injection 
```

Comment:
Well, SPIMolarFraction is not the molar mass, it is the fraction of the various pellet shards which consists of the ion species being created here. So I think it is very SPI-specific?

Of course I could have an additional flag instead of setting it to -1, but I thought it would be easier this way, while still not causing any harm?

If I did set it to None, how would it appear in the c++ code then? Anything would work as long as it is easy to check in the c++ code wether this species is a part of any SPI shards. Also, when I have had other parameters set to None in the final settings, I have got a DREAMIOException, so I'm not sure it would work?

Reply:
Ah, I see. Makes sense! I was a bit confused about how the parameter was used in the code, but I think I understand what it does now. If you the parameter is an array, and *some* elements can indicate that the species is disabled, then ``-1`` probably makes most sense. I typically handle ``None`` in the Python interface in the ``todict()`` method specifically and simply exclude the parameter there if it is ``None``. Then the kernel gets the default value automatically.

On a very minor, somewhat related, note: how do you check for ``-1`` in the kernel? If you don't already do it, it's probably safest to look for ``<0`` to avoid any error if conversion from different floating-point standards would occur.